### PR TITLE
[WIP] Strong Typing

### LINKED ISSUE #39: Use type-hints for all function arguments & return values
Since we're already using Python 3.6, we should consider also specifying all our input parameters with input & return types.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/48719674-use-type-hints-for-all-function-arguments-return-values?utm_campaign=plugin&utm_content=tracker%2F70027116&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F70027116&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by gronke**
closes #39 (Work in progress)

With merging this PR all interfaces in the library and cli modules will be strongly typed.

### CODE REVIEW

File: iocage/lib/Jail.py [hunk 1]
```diff
@@ -345,7 +345,10 @@ def start(
         self._run_hook("poststart")
 
     @property
-    def basejail_backend(self):
+    def basejail_backend(self) -> typing.Union[
+        iocage.lib.NullFSBasejailStorage.NullFSBasejailStorage,
+        iocage.lib.ZFSBasejailStorage.ZFSBasejailStorage
```
Comment (igalic) [inline iocage/lib/Jail.py]: maybe we should declare this union as its type somewhere…

File: iocage/lib/Jail.py [hunk 2]
```diff
@@ -1238,27 +1244,27 @@ def env(self):
         return jail_env
 
     @property
-    def identifier(self):
+    def identifier(self) -> str:
```
Comment (igalic) [inline iocage/lib/Jail.py]: IIRC, we discussed sometime, somewhere, that this and `name` need consistent (but different) validation.

File: iocage/lib/Jail.py [hunk 3]
```diff
@@ -1238,27 +1244,27 @@ def env(self):
...
+        return iocage.lib.Release.ReleaseGenerator(
             name=self.config["release"],
             logger=self.logger,
             host=self.host,
             zfs=self.zfs
         )
 
     @property
-    def logfile_path(self):
+    def logfile_path(self) -> str:
```
Comment (igalic) [inline iocage/lib/Jail.py]: la la #113

File: iocage/lib/Network.py [hunk 4]
```diff
@@ -30,13 +31,17 @@
...
-                 mtu=1500,
-                 bridges=None,
-                 logger=None):
+
+    def __init__(
+        self,
+        jail: 'iocage.lib.Jail.JailGenerator',
+        nic: str="vnet0",
+        ipv4_addresses: typing.Optional[typing.List[str]]=None,
+        ipv6_addresses: typing.Optional[typing.List[str]]=None,
```
Comment (igalic) [inline iocage/lib/Network.py]: do we have validation for these anywhere?

**Comment by gronke** [comment]
This PR is outdated and the interfaces that were enhanced here have already been typed and documented.

**Pull request closed without merging**